google-cloud-bigquerystorage/src/main/java/com/google/cloud/bigquery/storage/v1alpha2/BQTableSchemaToProtoDescriptor.java

```diff
@@ -0,0 +1,145 @@
+   * @param scope Keeps track of current scope to prevent repeated naming while constructing
+   *     descriptor.
+   * @throws Descriptors.DescriptorValidationException
+   */
+  private static Descriptor ConvertBQTableSchemaToProtoDescriptorImpl(
+      Table.TableSchema BQTableSchema, String scope)
+      throws Descriptors.DescriptorValidationException {
+    List<FileDescriptor> dependenciesList = new ArrayList<FileDescriptor>();
+    List<FieldDescriptorProto> fields = new ArrayList<FieldDescriptorProto>();
+    int index = 1;
+    for (Table.TableFieldSchema BQTableField : BQTableSchema.getFieldsList()) {
+      if (BQTableField.getType() == Table.TableFieldSchema.Type.STRUCT) {
+        String currentScope = scope + BQTableField.getName();
+        dependenciesList.add(
+            ConvertBQTableSchemaToProtoDescriptorImpl(
+                    Table.TableSchema.newBuilder()
+                        .addAllFields(BQTableField.getFieldsList())
+                        .build(),
+                    currentScope)
+                .getFile());
+        fields.add(ConvertBQStructToProtoMessage(BQTableField, index++, currentScope));
+      } else {
+        fields.add(ConvertBQTableFieldToProtoField(BQTableField, index++));
+      }
+    }
+    FileDescriptor[] dependenciesArray = new FileDescriptor[dependenciesList.size()];
+    dependenciesArray = dependenciesList.toArray(dependenciesArray);
+    DescriptorProto descriptorProto =
+        DescriptorProto.newBuilder().setName(scope).addAllField(fields).build();
+    FileDescriptorProto fileDescriptorProto =
+        FileDescriptorProto.newBuilder().addMessageType(descriptorProto).build();
+    FileDescriptor fileDescriptor =
+        FileDescriptor.buildFrom(fileDescriptorProto, dependenciesArray);
+    Descriptor descriptor = fileDescriptor.findMessageTypeByName(scope);
+    return descriptor;
+  }
+
+  /**
+   * Constructs a FieldDescriptorProto for non-struct BQ fields.
+   *
+   * @param BQTableField BQ Field used to construct a FieldDescriptorProto
+   * @param index Index for protobuf fields.
+   */
+  private static FieldDescriptorProto ConvertBQTableFieldToProtoField(
+      Table.TableFieldSchema BQTableField, int index) {
+    String fieldName = BQTableField.getName();
+    Table.TableFieldSchema.Mode mode = BQTableField.getMode();
+    return FieldDescriptorProto.newBuilder()
+        .setName(fieldName)
+        .setType((FieldDescriptorProto.Type) BQTableSchemaTypeMap.get(BQTableField.getType()))
+        .setLabel((FieldDescriptorProto.Label) BQTableSchemaModeMap.get(mode))
+        .setNumber(index)
+        .build();
+  }
+
+  /**
+   * Constructs a FieldDescriptorProto for a Struct type BQ field.
+   *
+   * @param BQTableField BQ Field used to construct a FieldDescriptorProto
+   * @param index Index for protobuf fields.
+   * @param scope Need scope to prevent naming issues
+   */
+  private static FieldDescriptorProto ConvertBQStructToProtoMessage(
```

Comment:
This function seems pretty overlapped with ConvertBQTableFieldToProtoField above.
Can we merge this two function into one?

Reply:
Sure!